### ISSUE TITLE
Agent may download a malicious executable if server is compromised

### ISSUE THREAD

**Issue by tosiara** [s3inlc/hashtopolis#581]
Currently the agent will download and run any executable that server instructs it in the response. In case the server gets compromised all the agents will get infected too very easily:

1. register a malicious cracker
2. create a task with the new cracker
3. assign the task to all the agents
4. agents will download and run the malicious executable 

I would suggest the agent to accept only whitelisted hashes of 7z archives. So user can choose during agent installation:

1. secure mode - provide trusted SHA512 hashes of 7z archives
2. old insecure mode - disable any checks and run whatever server returns

In this case user is explicitly confirms that the risk is known to him and he is responsible for any damage in case server gets hacked

**Comment by blazer2x** [comment]
The file hash method won't stop the compromised server from issuing a comprised agent which ignores the hash checking though. Will need something better.

**Comment by tosiara** [comment]
Sorry, I didn't get your comment

If SHA512 is written into agent's local config during installation then:

1. agenet asks server for a task
2. server assigns a malicious task with a new cracker
3. agent downloads cracker's 7z from a spacified URL
4. agent verifies SHA512 of a 7z file and rejects it without even extracting

**Comment by blazer2x** [comment]
The server can issue a fake update to the agent, with a modified agent which will override the safety check if the server is compromised.

**Comment by tosiara** [comment]
I didn't know there is a way to update the agent remotely
Good that my agent ZIP is owned by different user than pyhton process is running and thus is read-only. I hope this is enough to prevent remote modification of the agent. But I will verify this scenario too, thanks for pointing

**Comment by H4xl0r** [comment]
> 
> 
> I didn't know there is a way to update the agent remotely
> Good that my agent ZIP is owned by different user than pyhton process is running and thus is read-only. I hope this is enough to prevent remote modification of the agent. But I will verify this scenario too, thanks for pointing

he is on the right and easiest way to implent more sec

**Comment by roycewilliams** [comment]
I honestly can't tell what a better way to manage this would be. Even code-signing certs would have to be on the server itself in order to sign new clients, which an attacker who compromised the server would have access to. So you'd have to have a process where the server administrator keeps a code-signing cert offline somehow. I don't see that this is worth the trade-off.

**Comment by evilmog** [comment]
Honestly by forcing whitelisting you would kill the dynamic capabilities of the system. By installing hashtopolis you are basically making the hashtopolis master the admin of all your systems. Its really designed for a closed network or with machines you trust and control.

I could see hashing all the files but you need an inband or out of band signaling system, I do my sync over rsync so ymmv but really protect your hashtopolis master and assume if its popped then every sub system is popped including your cracking nodes

**Comment by tosiara** [comment]
> By installing hashtopolis you are basically making the hashtopolis master the admin of all your systems

I disagree with this statement. Personally, by installing hashtopolis agent I only give permission to GPU, and that's it. Think about it as a client option similar to `Trust agent with secret data` which is not on by default on the server side

So it would be nice to allow user explicitly select during agent install: does he want to grant full admin rights to the server or only allow access GPU? This would be very important feature

**Comment by TheWorkingDeveloper** [comment]
By running as a process on the system it has a level of access. That's it. If you don't run the program as admin then it won't gain admin rights but it will still run as the user you run the software as. And as a distributed system program there's fairly little you can do about this. The best you can do is set up a separate system/user/environment for your hashtopolis client and attach the GPU's directly. Such as a VM or K8 with passthrough. 

All in all, you have to generally accept the risk to my knowledge. My suggestion is to have a virus scanner / anomaly detection system on all systems to catch anything out of the ordinary.

**Comment by Nag1bat0r** [comment]
Add a script that compare checksum of hashtopolis hashcat with original hashcat from offsite each time. 
Client dowload this file manualy and it can not by auto updated. It checksumm is diffrent - "ALARM!!!" and stop executing next string (python3 hashtopolis.zip)

**Comment by evilmog** [comment]
I like that route

**Comment by Nag1bat0r** [comment]
"add an option to the client"

If i am a EVIL owner of hashtopolis server - i am an owner of the client (hashtopolis.zip) and i can modify it to bypass any protection...

**Comment by roycewilliams** [comment]
@Nag1bat0r There is usually no requirement that the client be one provided by the server. Both hashcat and the hashtopolis client can be indpendently downloaded and usually Just Work.

Of course, since the entire Hashtopolis model is to pass commands to hashcat on the command line, no amount of client selection hygiene is going to mitigate the fundamental "hey, please run this arbitrary commandline for me" model. :D

But it would at least eliminate the direct-executable tier of the problem.

**Comment by ObsidianOracle** [comment]
Feel free to implement that. Won´t be adressed by the core open-source dev team.
